zrml/market-commons/src/tests.rs

```diff
@@ -0,0 +1,317 @@
+        assert_ok!(MarketCommons::push_market(market_mock(2)));
+        assert_eq!(MarketCommons::market(&0).unwrap().oracle, 0);
+        assert_eq!(MarketCommons::market(&1).unwrap().oracle, 1);
+        assert_eq!(MarketCommons::market(&2).unwrap().oracle, 2);
+    });
+}
+
+#[test]
+fn markets_interacts_correctly_with_push_market() {
+    ExtBuilder::default().build().execute_with(|| {
+        assert_ok!(MarketCommons::push_market(market_mock(0)));
+        assert_ok!(MarketCommons::push_market(market_mock(1)));
+        assert_ok!(MarketCommons::push_market(market_mock(2)));
+        assert_eq!(<Markets<Runtime>>::get(0).unwrap().oracle, 0);
+        assert_eq!(<Markets<Runtime>>::get(1).unwrap().oracle, 1);
+        assert_eq!(<Markets<Runtime>>::get(2).unwrap().oracle, 2);
+    });
+}
+
+#[test]
+fn market_fails_if_market_does_not_exist() {
+    ExtBuilder::default().build().execute_with(|| {
+        assert_noop!(MarketCommons::market(&0), crate::Error::<Runtime>::MarketDoesNotExist);
+        assert_ok!(MarketCommons::push_market(market_mock(0)));
+        assert_ok!(MarketCommons::push_market(market_mock(1)));
+        assert_ok!(MarketCommons::push_market(market_mock(2)));
+        assert_noop!(MarketCommons::market(&3), crate::Error::<Runtime>::MarketDoesNotExist);
+    });
+}
+
+#[test]
+fn mutate_market_succeeds_if_closure_succeeds() {
+    ExtBuilder::default().build().execute_with(|| {
+        assert_ok!(MarketCommons::push_market(market_mock(0)));
+        assert_ok!(
+            // We change the market to check that `mutate_market` is actually no-op.
+            MarketCommons::mutate_market(&0, |market| {
+                market.oracle = 1;
+                Ok(())
+            })
+        );
+        assert_eq!(MarketCommons::market(&0).unwrap().oracle, 1);
+    });
+}
+
+#[test]
+fn mutate_market_fails_if_market_does_not_exist() {
+    ExtBuilder::default().build().execute_with(|| {
+        assert_noop!(
+            MarketCommons::mutate_market(&0, |_| Ok(())),
+            crate::Error::<Runtime>::MarketDoesNotExist
+        );
+        assert_ok!(MarketCommons::push_market(market_mock(0)));
+        assert_ok!(MarketCommons::push_market(market_mock(1)));
+        assert_ok!(MarketCommons::push_market(market_mock(2)));
+        assert_noop!(
+            MarketCommons::mutate_market(&3, |_| Ok(())),
+            crate::Error::<Runtime>::MarketDoesNotExist
+        );
+    });
+}
+
+#[test]
+fn mutate_market_is_noop_if_closure_fails() {
+    ExtBuilder::default().build().execute_with(|| {
+        let err = DispatchError::Other("foo");
+        assert_ok!(MarketCommons::push_market(market_mock(0)));
+        assert_noop!(
+            // We change the market to check that `mutate_market` is actually no-op.
+            MarketCommons::mutate_market(&0, |market| {
+                market.oracle = 1;
+                Err(err)
+            }),
+            err
+        );
+    });
```

Comment:
This isn't necessary. `assert_noop` will check that the on-chain memory remains unchanged. Otherwise, we would have to verify memory everytime we test for an error.